### PR TITLE
`Exam mode`: Fix problems with highlighting of the differences in the problem statement

### PR DESCRIPTION
**Pull request by coolchock**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Currently there is an issue that differences in a problem statement are not properly highlighted when a change occurs inside of code tag (single backtick operator) or in the task that gets injected into the markdown.
This pull requests resolves this issue and ensures that differences are rendered correctly, regardless of their location in the markdown.

### Description
Previously, differences between two problem statements were calculated at the markdown level and then markdown with highlighted differences was rendered. This occasionally led to errors, especially when changes occurred within special syntax such as code blocks or URLs. With the new approach, the problem statement markdowns are first converted into HTML. Then, differences between rendered problem statements are calculated using an HTML diff library.

The logic to remove tasks and inject them back into the document was moved to `ProgrammingExerciseInstructionComponent` to enable calculating tasks from the markdown with differences.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a programming exercise

There is already an exam called 'problem-statement-updates' in 'Test Course Michal Kawka' on ts2. Test users 1-7 are registered for the exam.

1. Log in to Artemis as a student in one browser
2. Log in to Artemis as an instructor in the second browser or incognito tab
3. Navigate to the exam as a student and open the programming exercise
4. As instructor, make some changes to the problem statement inside inside of special syntax (triple backtip, single backtick, <>, etc.)
5. As student, verify that the differences are highlighted correctly
6. As instructor, make some changes to the problem statement in one of the tasks. Task looks as follows (this is not highlighting, just an explanation what is meant by task):
image

7. As student, verify that the differences are highlighted correctly
8. Please also make some changes in the other places, for example headers, bold and italic parts of the markdown and ensure that everything works as expected.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Screenshots

Differences inside single backticks
Before:
image
After:
image

Differences in the tasks
Before:
image
After:
image

## Summary by CodeRabbit

- **New Features**
  - Enhanced problem statement highlighting in exam exercises using HTML for better visual differentiation.
  - Improved markdown to HTML conversion for more accurate problem statement displays.

- **Updates**
  - All exam submission components now display problem statements using an HTML variable for consistency and clarity.
  - Problem statement update methods now accept HTML inputs for uniform handling across different exam exercise types.

### CODE REVIEW

**Comment by pzdr7** [comment]
It could be out of scope for this PR, but URLs have a similar problem:

I changed

```md
[url](https://github.com/ls1intum/Artemis/pull/8461)
```

to

```md
[url](https://github.com/ls1intum/Artemis/pull/8462)
```

and got

image

**Comment by krusche** [comment]
I'm not sure if escaping is a good solution here. There will always be difficulties for special syntax. Would it not make more sense to use some kind of html diff?

File: src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts [hunk 1]
```diff
@@ -312,6 +309,39 @@ export class ProgrammingExerciseInstructionComponent implements OnChanges, OnDes
...
+            // Differences between UMLs are ignored, and we only inject the current one
+            setTimeout(() => {
+                const injectUML = this.injectableContentForMarkdownCallbacks[this.injectableContentForMarkdownCallbacks.length - 1];
+                if (injectUML) {
+                    injectUML();
+                }
+            }, 0);
+        } else {
+            this.injectableContentForMarkdownCallbacks = [];
+            this.renderedMarkdown = this.markdownService.safeHtmlForMarkdown(this.exercise.problemStatement, this.markdownExtensions);
```
Comment (pzdr7) [inline src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts:334]: I had an error in my console here. It didn't cause any problems, but can you double-check if `this.exercise` is ever `undefined` when this method is called?

It seems to happen when I first open the problem statement of a programming exercise.

image

**Review by pzdr7 — COMMENTED** [review]
I think the changes work rather well now. I tested it on TS2.

----

Things I've noticed:

- For a URL, if only the link changes, there's no visible diff.

But I don't know how relevant this edge case is & I think this might be hard to resolve with the HTML diff.

- Programming: When viewing the diff, I cannot click on the test cases for feedback. It works when I view only the new version.

image

**Review by BBesrour — DISMISSED** [review]
Tested in a testing session. Other than the issues mentioned by previous reviewers, works as intended

**Review by BaumiCoder — DISMISSED** [review]
Works so far, with the already mentioned issues. 
In Addition, a line break in text, which have no impact on the HTML output, is show as a deleted whitespace, but I think that is no show blocker.

File: src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts (hunk 1 above)
Comment (coolchock) [reply src/main/webapp/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts]: Regarding URLs, I don't think there's something we can do about it. I can open an issue on the library's github and maybe they'll find a way to handle this case in next versions.

When it comes to the feedback, I managed to make it clickable again, both for new version and version with diffs.

I also added check for undefined, the error shouldn't appear in the console anymore.

**Review by DominikRemo — DISMISSED** [review]
Tested in testing session on ts1. LGTM

**Review by kaancayli — APPROVED** [review]
Tested during testing session on TS1. Works as expected. LGTM 👍

**Review by SimonEntholzer — APPROVED** [review]
Tested in testing session

**Review by EneaGore — DISMISSED** [review]
Tested in testing session. Lgtm

**Review by BBesrour — APPROVED** [review]
tested on ts1, works fine

File: src/test/javascript/spec/component/exam/participate/exam-exercise-update-highlighter.component.spec.ts [hunk 2]
```diff
@@ -82,17 +79,16 @@ describe('ExamExerciseUpdateHighlighterComponent', () => {
...
                 });
         });
 
-        it('should ignore plantuml diagrams in programming exercise problem statements for diff calculation', () => {
-            const result = component.exercise.problemStatement;
-            expect(result).toEqual(component.updatedProblemStatementWithHighlightedDifferences);
-            fixture.detectChanges();
+        it('should not highlight differences for programming exercise', () => {
+            expect(component.updatedProblemStatementWithHighlightedDifferencesHTML).toBeUndefined();
+            expect(component.updatedProblemStatementHTML).toBeUndefined();
```
Comment (pzdr7) [inline src/test/javascript/spec/component/exam/participate/exam-exercise-update-highlighter.component.spec.ts:91]: This test case is not super intuitive for me. Could you explain what's going on here?

**Review by pzdr7 — DISMISSED** [review]
Code looks good. One small question

**Review by krusche — APPROVED** [review]
